framework/ui/internal/uiengine.cpp

```diff
@@ -95,6 +96,25 @@ void UiEngine::setup(QQmlEngine* e)
 #endif
 
     m_engine->addImportPath(":/qml");
+
+#ifdef QML_LOAD_FROM_SOURCE
+    for (const QString& path : m_sourceImportPaths) {
+        m_engine->addImportPath(path);
+    }
+#endif
+}
+
+void UiEngine::addSourceImportPath(const QString& path)
+{
+#ifdef QML_LOAD_FROM_SOURCE
+    LOGD() << path;
+    m_sourceImportPaths << path;
+    if (m_engine) {
+        m_engine->addImportPath(path);
```

Comment:
Can we only save paths here, without adding paths to the engine?
Because now in each module we add the import path to the engine, and then we call the setup method in appShell, where we add them all again.

Reply:
if there is no engine (this is null), then the setup was not called. If the engine is not null, then the setup has already been called.